google/cloud/bigtable/admin/internal/wait_for_consistency.cc

```diff
@@ -0,0 +1,184 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/bigtable/admin/internal/wait_for_consistency.h"
+#include "google/cloud/bigtable/admin/bigtable_table_admin_client.h"
+#include "google/cloud/bigtable/admin/bigtable_table_admin_options.h"
+#include "google/cloud/bigtable/admin/internal/bigtable_table_admin_option_defaults.h"
+#include <chrono>
+
+namespace google {
+namespace cloud {
+namespace bigtable_admin_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using RespType = StatusOr<bigtable::admin::v2::CheckConsistencyResponse>;
+using TimerFuture = future<StatusOr<std::chrono::system_clock::time_point>>;
+
+// This class borrows heavily from `google::cloud::internal::AsyncRetryLoop`
+class AsyncWaitForConsistencyImpl
+    : public std::enable_shared_from_this<AsyncWaitForConsistencyImpl> {
+ public:
+  AsyncWaitForConsistencyImpl(CompletionQueue cq,
+                              bigtable_admin::BigtableTableAdminClient client,
+                              std::string table_name,
+                              std::string consistency_token,
+                              Options const& options)
+      : cq_(std::move(cq)),
+        client_(std::move(client)),
+        table_name_(std::move(table_name)),
+        consistency_token_(std::move(consistency_token)),
+        options_(BigtableTableAdminDefaultOptions(std::move(options))),
+        polling_policy_(
+            options_
+                .get<bigtable_admin::BigtableTableAdminPollingPolicyOption>()) {
+  }
+
+  future<Status> Start() {
+    auto w = std::weak_ptr<AsyncWaitForConsistencyImpl>(shared_from_this());
+    result_ = promise<Status>([w] {
+      if (auto self = w.lock()) self->Cancel();
+    });
+
+    StartAttempt();
+    return result_.get_future();
+  }
+
+ private:
+  enum State {
+    kIdle,
+    kWaiting,
+    kDone,
+  };
+
+  void StartAttempt() {
+    auto self = this->shared_from_this();
+    auto op =
+        client_.AsyncCheckConsistency(table_name_, consistency_token_, options_)
+            .then([self](future<RespType> f) { self->OnAttempt(f.get()); });
+    SetWaiting(std::move(op));
```

Comment:
What happens if the `.then()` parameter is called before `SetWaiting()` gets a chance to run?